### PR TITLE
Nexus writer

### PR DESCRIPTION
**Pull request by Modularius**
A functional tool which creates a NeXus file upon the start of a run, and appends event data to it when frame-aggregated or digitizer event list messages are created.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,4 +1,3 @@
----
```
Comment (DanNixon) [inline .github/workflows/ci.yml:1]: Please revert the changes to this file.

File: nexus-writer/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "nexus-writer"
+version.workspace = true
+license.workspace = true
+edition.workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
```
Comment (DanNixon) [inline nexus-writer/Cargo.toml:8]:
```suggestion
```

File: nexus-writer/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,326 @@
+mod event_message;
+mod metrics;
+mod nexus;
+
+use anyhow::{anyhow, Result};
+use chrono::Duration;
+use clap::Parser;
+use event_message::GenericEventMessage;
+use kagiyama::{AlwaysReady, Watcher};
+use ndarray as _;
```
Comment (DanNixon) [inline nexus-writer/src/main.rs:10]: Is this actually used?

File: nexus-writer/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+use ndarray_stats as _;
```
Comment (DanNixon) [inline nexus-writer/src/main.rs:11]: Is this actually used?

File: nexus-writer/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt::init();
+
+    let args = Cli::parse();
+    debug!("Args: {:?}", args);
+
+    let mut watcher = Watcher::<AlwaysReady>::default();
+    metrics::register(&mut watcher);
+    //watcher.start_server(args.observability_address).await;
```
Comment (DanNixon) [inline nexus-writer/src/main.rs:81]: Is this commented out for a reason?

**Review by DanNixon — CHANGES_REQUESTED** [review]
Just an initial look through.

There are also a few thousand lines of generated files that have been committed, please remove those.

File: trace-reader/src/main.rs [hunk 6]
```diff
@@ -48,6 +53,14 @@ struct Cli {
     #[clap(long, default_value = "1")]
     number_of_trace_events: usize,
 
+    /// Add this value to the channel ids
+    #[clap(long, default_value = "0")]
+    channel_id_shift: Channel,
```
Comment (DanNixon) [inline trace-reader/src/main.rs:58]:
```suggestion
    channel_id_offset: Channel,
```

File: trace-reader/src/processing.rs [hunk 7]
```diff
@@ -22,26 +22,31 @@ use supermusr_streaming_types::{
 };
 
 /// Reads the contents of trace_file and dispatches messages to the given Kafka topic.
+#[allow(clippy::too_many_arguments)]
```
Comment (DanNixon) [inline trace-reader/src/processing.rs:25]: No action is needed immediately, but if you find yourself adding suppressions like this it should be taken as a sign that there might be a design/architecture issue.

File: nexus-writer/src/nexus/builder.rs [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        assert!(nexus.start_command(start2).is_err());
+    }
+
+    #[test]
+    fn frame_messages_correct() {
+        let mut nexus = Nexus::new(None);
+        let mut fbb = FlatBufferBuilder::new();
+
+        let ts = GpsTime::new(0, 1, 0, 0, 16, 0, 0, 0);
+        let ts_start = {
```
Comment (DanNixon) [inline nexus-writer/src/nexus/builder.rs:204]: What is the purpose of the extra scope here?

File: nexus-writer/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    control_topic: String,
+
+    #[clap(long)]
+    digitiser_event_topic: Option<String>,
+
+    #[clap(long)]
+    frame_event_topic: Option<String>,
+
+    #[clap(long)]
+    histogram_topic: Option<String>,
```
Comment (DanNixon) [inline nexus-writer/src/main.rs:55]: We only need the option for this to be added in the future, it is not needed now.

File: nexus-writer/src/nexus/builder.rs [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+            "\nNexus Context: {0:?}\n{e}",
+            self.run_cache
+                .iter()
+                .map(|run| run.get_name())
+                .collect::<Vec<_>>()
+        )
+    }
+
+    #[cfg(test)]
+    fn cache_iter(&self) -> vec_deque::Iter<'_, Run> {
```
Comment (DanNixon) [inline nexus-writer/src/nexus/builder.rs:39]: Why is this needed?

File: nexus-writer/src/nexus/run_parameters.rs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+#[derive(Default, Debug)]
+pub(crate) struct RunStopParameters {
+    pub(crate) collect_until: u64,
+    pub(crate) last_modified: DateTime<Utc>,
+}
+
+#[derive(Debug)]
+pub(crate) struct RunParameters {
+    pub(crate) collect_from: u64,
```
Comment (DanNixon) [inline nexus-writer/src/nexus/run_parameters.rs:15]: Can we use a time type here.

File: nexus-writer/src/nexus/run_parameters.rs [hunk 12]
```diff
@@ -0,0 +1,72 @@
+use anyhow::{anyhow, Result};
+use chrono::{DateTime, Utc};
+use supermusr_streaming_types::{
+    ecs_6s4t_run_stop_generated::RunStop, ecs_pl72_run_start_generated::RunStart,
+};
+
+#[derive(Default, Debug)]
+pub(crate) struct RunStopParameters {
+    pub(crate) collect_until: u64,
```
Comment (DanNixon) [inline nexus-writer/src/nexus/run_parameters.rs:9]: Can we use a time type here.

File: nexus-writer/src/nexus/run_parameters.rs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    pub(crate) last_modified: DateTime<Utc>,
```
Comment (DanNixon) [inline nexus-writer/src/nexus/run_parameters.rs:10]: What is this used for?

File: nexus-writer/src/nexus/run.rs [hunk 14]
```diff
@@ -0,0 +1,76 @@
+use super::{hdf5::RunFile, RunParameters};
+use crate::event_message::GenericEventMessage;
+use anyhow::Result;
+use chrono::{DateTime, Duration, Utc};
+use std::path::Path;
+use supermusr_streaming_types::ecs_6s4t_run_stop_generated::RunStop;
+
+#[derive(Debug)]
+pub(crate) struct Run {
+    parameters: RunParameters,
```
Comment (DanNixon) [inline nexus-writer/src/nexus/run.rs:10]: Is there a reason for storing all the parameters?

File: nexus-writer/src/metrics.rs [hunk 15]
```diff
@@ -0,0 +1,29 @@
+use kagiyama::{
```
Comment (DanNixon) [inline nexus-writer/src/metrics.rs:1]: Remove this metrics implementation, it will be replaced in the future.

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: nexus-writer/src/nexus/builder.rs (hunk 8 above)
Comment (Modularius) [reply nexus-writer/src/nexus/builder.rs:204]: I think it was a relic from an earlier hack, that I don't need anymore. Removed.

File: nexus-writer/src/nexus/run_parameters.rs (hunk 13 above)
Comment (Modularius) [reply nexus-writer/src/nexus/run_parameters.rs:10]: This is used in the run buffering feature. If messages are coming in slow, the run isn't flushed from cache until a set time has passed since it was last added to.

File: nexus-writer/src/nexus/run.rs (hunk 14 above)
Comment (Modularius) [reply nexus-writer/src/nexus/run.rs:10]: Possibly, not but I wanted to keep the flatbuffers code isolated from the hdf5 code as much as possible. So the RunParameters struct is the only interface between the two. I could place some of the run parameters into a new "RunHeader" object which would not be persisted. Is this worth doing do you think?

File: nexus-writer/src/nexus/builder.rs (hunk 10 above)
Comment (Modularius) [reply nexus-writer/src/nexus/builder.rs:39]: This is used in tests to check runs have been successfully created.

**Comment by DanNixon** [comment]
@Modularius why has this been closed?

File: nexus-writer/diagrams/.gitignore [hunk 16]
```diff
@@ -0,0 +1 @@
+!*.svg
```
Comment (DanNixon) [inline nexus-writer/diagrams/.gitignore:1]: Why was SVG ever added to the root gitignore?

File: nexus-writer/Readme.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+A timer is set to tick on intervals of `cache-poll-interval-ms`.
+On each tick, any run in memory is removed if:
+    - it has been terminated, and
+    - at least `cache-run-ttl-ms` of time has passed since the time in its `last_modified` field,
+This allows event-list messages to catch up with its run if it has been delayed in the pipeline.
+
+### Sequence Diagrams
+
+![Run Start](diagrams/RunStart.svg)
```
Comment (DanNixon) [inline nexus-writer/Readme.md:73]: Could these appear at relevant parts of the Behavior section.

File: nexus-writer/diagrams/CachePollIntervalTick.svg
Comment (DanNixon) [inline nexus-writer/diagrams/CachePollIntervalTick.svg:1]: Could these go in a directory named `docs` rather than `diagrams`.

File: nexus-writer/Readme.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
+# nexus-writer
```
Comment (DanNixon) [inline nexus-writer/Readme.md:1]: Looks like you need to rebase this PR, the documentation linting is not being picked up for some reason.

**Review by DanNixon — CHANGES_REQUESTED** [review]

**Comment by DanNixon** [comment]
@Modularius There are now a lot of irrelevant changes in the GitHub diff view, could you squash your commits locally and push a single commit.

**Comment by Modularius** [comment]
Yeah, sorry, I'll try that again.

**Comment by Modularius** [comment]
Would it be easier to create a new branch from main, copy over the nexus-writer files and use that for the PR?

**Comment by DanNixon** [comment]
> Would it be easier to create a new branch from main, copy over the nexus-writer files and use that for the PR?

If that works easier. The equivalent of that is the "Alternative Approach" listed in the link above.
